src/mediapipe_calculators/python_executor_calculator.cc

```diff
@@ -0,0 +1,160 @@
+//*****************************************************************************
+// Copyright 2023 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//*****************************************************************************
+#include <unordered_map>
+#include "../python/ovms_py_tensor.hpp"
+#include "../mediapipe_internal/pythonnoderesource.hpp"
+
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wdeprecated-declarations"
+#include "mediapipe/framework/calculator_framework.h"
+#pragma GCC diagnostic pop
+
+#include <Python.h>
+#include <pybind11/embed.h> // everything needed for embedding
+#include <pybind11/stl.h>
+
+#include "../python/python_backend.hpp"
+
+namespace py = pybind11;
+using namespace py::literals;
+using namespace ovms;
+
+namespace mediapipe {
+
+typedef std::unordered_map<std::string, std::shared_ptr<PythonNodeResource>> PythonNodesResources;
+
+const std::string INPUT_SIDE_PACKET_TAG = "PYTHON_NODE_RESOURCES";
+
+typedef std::unique_ptr<OvmsPyTensor> OvmsPyTensorPtr;
+
+class PythonExecutorCalculator : public CalculatorBase {
+
+/* 
+TODO: Streaming support:
+    - timestamping
+    - loopback input
+*/ 
+
+std::shared_ptr<PythonNodeResource> nodeResources;
+std::unique_ptr<PyObjectWrapper<py::iterator>> pyIteratorPtr;
+
+void pushOutputs(CalculatorContext* cc, py::list pyOutputs) {
+    py::gil_scoped_acquire acquire;
+    for (const std::string& tag : cc->Outputs().GetTags()) {
+        for (py::handle pyOutputHandle : pyOutputs) {
```

Comment:
Instead of having 2 for loops, we could push outputs in reversed manner:
1. First iterate over pyOutputs
2. For each pyOutput check its name and try to find its tag in cc (context). cc->Outputs().Tag(name).Add()

Reply:
Agreed